test/typescript/provider-upgrade-command/test.ts

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) HashiCorp, Inc
+// SPDX-License-Identifier: MPL-2.0
+import {
+  TestDriver,
+  onPosix,
+  onWindows,
+  packageJsonWithDependency,
+} from "../../test-helper";
+
+describe("provider upgrade command", () => {
+  let driver: TestDriver;
+  beforeEach(async () => {
+    driver = new TestDriver(__dirname, {
+      CDKTF_DIST: "",
+      DISABLE_VERSION_CHECK: "true",
+      CI: "1",
+    }); // reset CDKTF_DIST set by run-against-dist script & disable version check as we have to use an older version of cdktf-cli
+    await driver.setupTypescriptProject({
+      init: { additionalOptions: "--cdktf-version 0.10.4" },
+    });
+  }, 500_000);
+
+  describe("pre-built", () => {
+    beforeEach(async () => {
+      await driver.exec("cdktf", [
+        "provider",
+        "add",
+        "random@=3.1.3", // this is not the latest version, but theres v0.2.55 of the pre-built provider resulting in exactly this package
+      ]);
+    });
+
+    it("can update withing the same cdktf version to a specific version", async () => {
+      expect(driver.packageJson()).toEqual(
+        packageJsonWithDependency("@cdktf/provider-random", "^0.2.55")
+      );
+
+      await driver.exec("cdktf", ["provider", "upgrade", "random@=3.2.0"]);
+
+      expect(driver.packageJson()).toEqual(
+        packageJsonWithDependency("@cdktf/provider-random", "^0.2.64")
+      );
+    });
+
+    it("can update within the same cdktf version to the latest version", async () => {
+      expect(driver.packageJson()).toEqual(
+        packageJsonWithDependency("@cdktf/provider-random", "^0.2.55")
+      );
+
+      await driver.exec("cdktf", ["provider", "upgrade", "random"]);
+
+      // Assert that we have version 0.2.64
+      expect(driver.packageJson()).toEqual(
+        packageJsonWithDependency("@cdktf/provider-random", "^0.2.64")
+      );
+    });
+
+    it("can update withing the same cdktf version to the latest version in yarn", async () => {
+      // Pin random provider version so that the upgrade can do anything
+      await driver.exec("yarn", ["add", "@cdktf/provider-random@0.2.55"]);
+      expect(driver.packageJson()).toEqual(
+        packageJsonWithDependency("@cdktf/provider-random", "0.2.55")
+      );
+      await driver.exec("rm", ["-rf", "node_modules"]);
+      await driver.exec("rm", ["package-lock.json"]);
```

Comment:
Should this be `yarn.lock`? or should lines 65 & 59 also need changing from `yarn` to `npm`?

Reply:
No, `package-lock.json` is correct. The normal project setup installs dependencies through npm, we are cleaning up all npm artifacts and install with yarn to simulate a yarn project